keras_nlp/tests/test_case.py

```diff
@@ -223,3 +262,158 @@ def run_class_serialization_test(self, instance):
+
+    def run_task_test(
+        self,
+        cls,
+        init_kwargs,
+        train_data,
+        expected_output_shape=None,
+        batch_size=2,
+    ):
+        """Run basic tests for a backbone, including compilation."""
+        task = cls(**init_kwargs)
+        # Check serialization (without a full save).
+        self.run_serialization_test(task)
+        preprocessor = task.preprocessor
+        ds = tf.data.Dataset.from_tensor_slices(train_data).batch(batch_size)
+        x, y, sw = keras.utils.unpack_x_y_sample_weight(train_data)
+
+        # Test predict.
+        output = task.predict(x)
+        if expected_output_shape is not None:
+            output_shape = tree.map_structure(lambda x: x.shape, output)
+            self.assertAllClose(output_shape, expected_output_shape)
+        # With a dataset.
+        output_ds = task.predict(ds)
+        self.assertAllClose(output, output_ds)
+        # With split preprocessing.
+        task.preprocessor = None
+        output_split = task.predict(ds.map(preprocessor))
+        task.preprocessor = preprocessor
+        self.assertAllClose(output, output_split)
+
+        # Test fit.
+        task.fit(x, y, sample_weight=sw)
+        # With a dataset.
+        task.fit(ds)
+        # With split preprocessing.
+        task.preprocessor = None
+        task.fit(ds.map(preprocessor))
+        task.preprocessor = preprocessor
+
+    def run_preset_test(
+        self,
+        cls,
+        preset,
+        input_data,
+        init_kwargs={},
+        expected_output=None,
+        expected_output_shape=None,
+        expected_partial_output=None,
+    ):
+        """Run instantiation and a forward pass for a preset."""
+        self.assertRegex(cls.from_preset.__doc__, preset)
+
+        with self.assertRaises(ValueError):
+            cls.from_preset("clowntown", **init_kwargs)
+
+        instance = cls.from_preset(preset, **init_kwargs)
+
+        if isinstance(input_data, tuple):
+            # Mimic tf.data unpacking behavior for preprocessing layers.
+            output = instance(*input_data)
+        else:
+            output = instance(input_data)
+
+        if isinstance(instance, keras.Model):
+            instance = cls.from_preset(
+                preset, load_weights=False, **init_kwargs
+            )
+            instance(input_data)
+
+        if expected_output is not None:
+            self.assertAllClose(output, expected_output)
+
+        if expected_output_shape is not None:
+            output_shape = tree.map_structure(lambda x: x.shape, output)
+            self.assertAllClose(output_shape, expected_output_shape)
+
+        if expected_partial_output is not None:
+            # Allow passing a partial output snippet of the last dimension.
+            # We want check stability, but the full output would be too long.
+            def compare(actual, expected):
+                for _ in range(len(actual.shape) - 1):
```

Comment:
This feels really awkward but after some thinking I can't figure out a better way to do this.

Reply:
maybe just `actual.reshape(-1)[:len(expected)]` would do?

Reply:
yeah that works, simplified